### PR TITLE
Share settings for sole-tenant node groups.

### PR DESCRIPTION
**Pull request by idyczko**
This PR introduces support for cross-project sharing of Node Groups.
This PR resolves [Issue 13281](https://github.com/hashicorp/terraform-provider-google/issues/13281).

If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [X] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [X] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Note on acceptance tests**
I ran the acceptance tests locally with my own projects instead of automatically created projects due to some limitations in my organisation.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
compute: added `share_settings` field to the `google_compute_node_group` resource.
```

### CODE REVIEW

**Comment by idyczko** [comment]
I also wasn't sure what the TF policy is regarding resource validations. For instance, we duplicate some resource validations from the GCE REST API on the gcloud side but not others - please advise how to decide if a resource validation should be duplicated on the TF side.

**Comment by zli82016** [comment]
Added @c2thorn as the secondary reviewer.

**Comment by idyczko** [comment]
@zli82016 please let me know when I'm supposed to take any action here. I'm having trouble interpreting the results of the failed generate-diffs check. I'm getting
"You do not have sufficient permissions to view this page" on the GCP build page.

**Comment by zli82016** [comment]
> @zli82016 please let me know when I'm supposed to take any action here. I'm having trouble interpreting the results of the failed generate-diffs check. I'm getting "You do not have sufficient permissions to view this page" on the GCP build page.

@idyczko , I will check with the team what permission you need. Also, can you check my comments? Especially for the file [mmv1/templates/terraform/examples/node_group_share_settings.tf.erb]

**Comment by zli82016** [comment]
@idyczko , can you open this page https://pantheon.corp.google.com/cloud-build/builds/e8f13117-bee8-4059-8256-6ba814cb48ad?project=graphite-docker-images&mods=monitoring_api_prod ?

The generate-diffs failed because that `:test_env_vars` are used in file node_group_share_settings.tf.erb but not defined in terraform.yaml file.

**Review by zli82016 — CHANGES_REQUESTED** [review]

File: mmv1/products/compute/api.yaml
Comment (idyczko) [reply mmv1/products/compute/api.yaml:8930]: Thank you! Done.

File: mmv1/products/compute/terraform.yaml
Comment (idyczko) [reply mmv1/products/compute/terraform.yaml]: I copied it from the shareSettings already used by the Reservation resource. If you say it's not needed, let's get rid of it. Thanks!

File: mmv1/products/compute/terraform.yaml
Comment (idyczko) [reply mmv1/products/compute/terraform.yaml]: That was copied from the shareSettings used by the Reservation resource as well. Removed.

File: mmv1/products/compute/terraform.yaml
Comment (idyczko) [reply mmv1/products/compute/terraform.yaml]: Done.

File: mmv1/templates/terraform/examples/node_group_share_settings.tf.erb
Comment (idyczko) [reply mmv1/templates/terraform/examples/node_group_share_settings.tf.erb:4]: Done, thank you!

File: mmv1/templates/terraform/examples/node_group_share_settings.tf.erb
Comment (zli82016) [inline mmv1/templates/terraform/examples/node_group_share_settings.tf.erb:8]: This line can be removed as we don't need specify project. The default project from the provider will be used.

File: mmv1/products/compute/terraform.yaml
Comment (zli82016) [reply mmv1/products/compute/terraform.yaml]: Please check the doc for default_from_api https://googlecloudplatform.github.io/magic-modules/docs/how-to/add-mmv1-resource/#terraformyaml. We should be careful to use default_from_api property, as it is a breaking change to remove it later and requires a major release.

File: mmv1/templates/terraform/examples/node_group_share_settings.tf.erb
Comment (idyczko) [reply mmv1/templates/terraform/examples/node_group_share_settings.tf.erb]: Done, thank you!

File: mmv1/templates/terraform/examples/node_group_share_settings.tf.erb
Comment (idyczko) [reply mmv1/templates/terraform/examples/node_group_share_settings.tf.erb]: Done.

**Comment by zli82016** [comment]
LGTM. Please wait for the review from @c2thorn .

File: mmv1/products/compute/api.yaml
Comment (c2thorn) [inline mmv1/products/compute/api.yaml:8933]: This seems to be implemented correctly in terms of the API, but I cannot understand the design. The `project_id` the only relevant information being passed here, why is it duplicated in both the key and value within a map? Am I missing something? Do you know if the API is setting up to have other values within the `projectConfig` nested object?

We rarely deviate from the API, but it seems like a frivolous user experience to have to set the project id twice. We could relatively simply implement custom code to take a list of project id's from the user and supply the API with the map in the correct format. From my perspective, this could prevent user error. More context would help me understand however

Comment (idyczko) [reply mmv1/products/compute/api.yaml:8933]: To be honest, I cannot defend this decision, but I don't think it's been designed while working on this particular feature. The semantics of sharing node groups across projects was derived from how reservations are shared (take a look at the **shareSettings** field of the **Reservation** resource). Is there anything we can do with that at this point?

Also, I've stumbled upon some custom code that seems to be responsible for **shareSettings** modifications when updating the **Reservation** resource:
mmv1/templates/terraform/update_encoder/reservation.go.erb
Could you let me know if that is something we need for the **NodeGroup** resource as well?

Comment (zli82016) [reply mmv1/products/compute/api.yaml:8933]: We don't need to add update_encoder for `shareSettings` in the resource `NodeGroup`. The resource NodeGroup has property `input: true`. That means that a new resource is created no matter what field has changed (from the code, node_template is an exception). 

https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/products/compute/api.yaml#L8794

Waiting for @c2thorn 's opinions.

Comment (c2thorn) [reply mmv1/products/compute/api.yaml:8933]: @idyczko thank you for pointing out the `shareSettings` field in Reservation. Considering that we've already gone forward with it there, we may as well keep consistent here.

@zli82016 is correct that a custom update encoder should not be needed since the resource cannot be updated in place at all. If this is no longer true, we can modify this in a separate PR.

**Review by c2thorn — APPROVED** [review]

**Review by zli82016 — APPROVED** [review]
LGTM

**Pull request merged**
